In [1]:
import os
os.environ["OPENBLAS_NUM_THREADS"] = "1"  # For implicit ALS

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
import pandas as pd
import numpy as np

from implicit.als import AlternatingLeastSquares

from rectools.metrics import Precision, Recall, MAP, MeanInvUserFreq, Serendipity, calc_metrics
from rectools.models import PopularModel, RandomModel, ImplicitALSWrapperModel
from rectools import Columns
from rectools.model_selection import TimeRangeSplitter
from rectools.dataset import Dataset
from rectools.models import ImplicitALSWrapperModel, LightFMWrapperModel

import matplotlib.pyplot as plt
import seaborn as sns

import matplotlib.pyplot as plt
from pathlib import Path
import typing as tp
from tqdm import tqdm

from lightfm import LightFM

from implicit.bpr import BayesianPersonalizedRanking

from implicit.lmf import LogisticMatrixFactorization

import optuna
from optuna.samplers import TPESampler

# Data

In [4]:
interactions = pd.read_csv("../data/kion_train/interactions.csv")
users = pd.read_csv("../data/kion_train/users.csv")
items = pd.read_csv("../data/kion_train/items.csv")

In [5]:
Columns.Datetime = 'last_watch_dt'

In [6]:
def display_df(df):
    return pd.concat([df.head(), df.tail()])

In [7]:
display_df(interactions)

user_id  item_id last_watch_dt  total_dur  watched_pct
0         176549     9506    2021-05-11       4250         72.0
1         699317     1659    2021-05-29       8317        100.0
2         656683     7107    2021-05-09         10          0.0
3         864613     7638    2021-07-05      14483        100.0
4         964868     9506    2021-04-30       6725        100.0
5476246   648596    12225    2021-08-13         76          0.0
5476247   546862     9673    2021-04-13       2308         49.0
5476248   697262    15297    2021-08-20      18307         63.0
5476249   384202    16197    2021-04-19       6203        100.0
5476250   319709     4436    2021-08-15       3921         45.0

# Prepare features

## Users

In [8]:
display_df(users)

user_id         age        income  sex  kids_flg
0        973171   age_25_34  income_60_90    М         1
1        962099   age_18_24  income_20_40    М         0
2       1047345   age_45_54  income_40_60    Ж         0
3        721985   age_45_54  income_20_40    Ж         0
4        704055   age_35_44  income_60_90    Ж         0
840192   339025  age_65_inf   income_0_20    Ж         0
840193   983617   age_18_24  income_20_40    Ж         1
840194   251008         NaN           NaN  NaN         0
840195   590706         NaN           NaN    Ж         0
840196   166555  age_65_inf  income_20_40    Ж         0

In [9]:
users.isnull().sum()

user_id         0
age         14095
income      14776
sex         13831
kids_flg        0
dtype: int64

In [10]:
users.dtypes

user_id      int64
age         object
income      object
sex         object
kids_flg     int64
dtype: object

In [11]:
users.fillna("Unknown", inplace=True)

In [12]:
users.nunique()

user_id     840197
age              7
income           7
sex              3
kids_flg         2
dtype: int64

In [13]:
users.isnull().sum()

user_id     0
age         0
income      0
sex         0
kids_flg    0
dtype: int64

In [14]:
features = {
    "users": ["age", "income", "sex"]
}

## Items 

In [15]:
display_df(items)

item_id content_type                 title      title_orig  \
0        10711         film        Поговори с ней  Hable con ella   
1         2508         film           Голые перцы    Search Party   
2        10716         film      Тактическая сила  Tactical Force   
3         7868         film                45 лет        45 Years   
4        16268         film  Все решает мгновение             NaN   
15958     6443       series         Полярный круг   Arctic Circle   
15959     2367       series               Надежда             NaN   
15960    10632       series                Сговор          Hassel   
15961     4538       series          Среди камней      Darklands    
15962     3206       series                  Гоша             NaN   

       release_year                                            genres  \
0            2002.0           драмы, зарубежные, детективы, мелодрамы   
1            2014.0                  зарубежные, приключения, комедии   
2            2011.0  криминал, зарубежные, триллеры, боевики, комедии   
3            2015.0                      драмы, зарубежные, мелодрамы   
4            1978.0                драмы, спорт, советские, мелодрамы   
15958        2018.0                         драмы, триллеры, криминал   
15959        2020.0                                    драмы, боевики   
15960        2017.0                         драмы, триллеры, криминал   
15961        2019.0                            драмы, спорт, криминал   
15962        2019.0                                           комедии   

                 countries  for_kids  age_rating   studios  \
0                  Испания       NaN        16.0       NaN   
1                      США       NaN        16.0       NaN   
2                   Канада       NaN        16.0       NaN   
3           Великобритания       NaN        16.0       NaN   
4                     СССР       NaN        12.0  Ленфильм   
15958  Финляндия, Германия       NaN        16.0       NaN   
15959               Россия       0.0        18.0       NaN   
15960               Россия       0.0        18.0       NaN   
15961               Россия       0.0        18.0       NaN   
15962               Россия       0.0        16.0       NaN   

                                   directors  \
0                           Педро Альмодовар   
1                             Скот Армстронг   
2                           Адам П. Калтраро   
3                                  Эндрю Хэй   
4                           Виктор Садовский   
15958                          Ханну Салонен   
15959                         Елена Хазанова   
15960  Эшреф Рейбрук, Амир Камдин, Эрик Эгер   
15961          Марк О’Коннор, Конор МакМахон   
15962                         Михаил Миронов   

                                                  actors  \
0      Адольфо Фернандес, Ана Фернандес, Дарио Гранди...   
1      Адам Палли, Брайан Хаски, Дж.Б. Смув, Джейсон ...   
2      Адриан Холмс, Даррен Шалави, Джерри Вассерман,...   
3      Александра Риддлстон-Барретт, Джеральдин Джейм...   
4      Александр Абдулов, Александр Демьяненко, Алекс...   
15958  Иина Куустонен, Максимилиан Брюкнер, Пихла Вии...   
15959  Виктория Исакова, Александр Кузьмин, Алексей М...   
15960  Ола Рапас, Алиетт Офейм, Уильма Лиден, Шанти Р...   
15961  Дэйн Уайт О’Хара, Томас Кэйн-Бирн, Джудит Родд...   
15962                 Мкртыч Арзуманян, Виктория Рунцова   

                                             description  \
0      Мелодрама легендарного Педро Альмодовара «Пого...   
1      Уморительная современная комедия на популярную...   
2      Профессиональный рестлер Стив Остин («Все или ...   
3      Шарлотта Рэмплинг, Том Кортни, Джеральдин Джей...   
4      Расчетливая чаровница из советского кинохита «...   
15958  Во время погони за браконьерами по лесу, сотру...   
15959  Оригинальный киносериал от создателей «Бывших»...   
15960  Криминальная драма по мотивам романов о шведск...   
15961  Семнадцатилетний Дэмиен

In [16]:
items.isnull().sum()

item_id             0
content_type        0
title               0
title_orig       4745
release_year       98
genres              0
countries          37
for_kids        15397
age_rating          2
studios         14898
directors        1509
actors           2619
description         2
keywords          423
dtype: int64

In [17]:
items.dtypes

item_id           int64
content_type     object
title            object
title_orig       object
release_year    float64
genres           object
countries        object
for_kids        float64
age_rating      float64
studios          object
directors        object
actors           object
description      object
keywords         object
dtype: object

In [18]:
items.nunique()

item_id         15963
content_type        2
title           15293
title_orig      10575
release_year      105
genres           2771
countries         687
for_kids            2
age_rating          6
studios            38
directors        7973
actors          12996
description     15619
keywords        15492
dtype: int64

In [19]:
items["countries"].fillna("Unknown", inplace=True)

In [20]:
items["genres"] = items["genres"].str.lower().str.replace(", ", ",", regex=False).str.split(",")
items["countries"] = items["countries"].str.lower().str.replace(", ", ",", regex=False).str.split(",")

In [21]:
features.update(
    {
        "items": ["content_type", "genres", "countries"]
    }
)

**Фичи**, которые **будут использоваться** для users-ов и item-ов соответственно

In [22]:
features

{'users': ['age', 'income', 'sex'],
 'items': ['content_type', 'genres', 'countries']}

In [23]:
interactions[Columns.Datetime] = pd.to_datetime(interactions[Columns.Datetime], format='%Y-%m-%d')

In [24]:
max_date = interactions[Columns.Datetime].max()
max_date

Timestamp('2021-08-22 00:00:00')

In [25]:
# Бинаризация процента просмотренного на 10 частей
interactions[Columns.Weight] = pd.qcut(
    interactions["watched_pct"], 10, duplicates = "drop"
).astype("category").cat.codes

# Create CV

В соответствии с ДЗ-3 **валидацию** сделаем по **7 дней**

In [26]:
def create_data_range(
    last_date: pd.Timestamp, 
    n_folds: int = 7, 
    unit: str = "W", 
    n_units: int = 1, 
    show: bool = True,
):
    periods = n_folds + 1
    freq = f"{n_units}{unit}"
    
    start_date = last_date - pd.Timedelta(n_folds * n_units + n_units, unit=unit)  
    
    date_range = pd.date_range(start=start_date, periods=periods, freq=freq, tz=last_date.tz)
    
    if show:
        print(
            f"start_date: {start_date}\n"
            f"last_date: {last_date}\n"
            f"periods: {periods}\n"
            f"freq: {freq}\n"
            f"Test fold borders: {date_range.values.astype('datetime64[D]')}\n"
        )
        
    return date_range

Создадим **три фолда** в силу ограниченности по времени

In [27]:
CONFIG_CV = {
    "cv_v1": {
        "n_folds": 3,
        "unit": "W",
        "n_units": 1,
    }, 
}

In [28]:
date_range = create_data_range(
    max_date, 
    n_folds=CONFIG_CV["cv_v1"]["n_folds"], 
    unit=CONFIG_CV["cv_v1"]["unit"], 
    n_units=CONFIG_CV["cv_v1"]["n_units"]
)

start_date: 2021-07-25 00:00:00
last_date: 2021-08-22 00:00:00
periods: 4
freq: 1W
Test fold borders: ['2021-07-25' '2021-08-01' '2021-08-08' '2021-08-15']



In [29]:
cv = TimeRangeSplitter(
    date_range=date_range,
    filter_already_seen=True,
    filter_cold_items=True,
    filter_cold_users=True,
)

In [30]:
from rectools.dataset import Interactions
print(f"Real number of folds: {cv.get_n_splits(Interactions(interactions))}")

Real number of folds: 3


# Create Metrics

In [31]:
metrics = {
    "prec@10": Precision(k=10),
    "recall@10": Recall(k=10),
    "MAP@10": MAP(k=10),
    "novelty": MeanInvUserFreq(k=10),
    "serendipity": Serendipity(k=10),
}

metrics

{'prec@10': Precision(k=10),
 'recall@10': Recall(k=10),
 'MAP@10': MAP(k=10, divide_by_k=False),
 'novelty': MeanInvUserFreq(k=10),
 'serendipity': Serendipity(k=10)}

# Models

**Модель на основе популряного возьмем из ДЗ-3**. 
Холодным юзерам будет возвращаться популярное

In [32]:
# hack from https://gist.github.com/crcrpar/c6d6fdf8112280654884353d6e68f6bb
suggested_params = {}


def create_features_dataset(col, train, df_value, features):
    """
        Create feature dataset for users / items
    """
    df_value = df_value.loc[df_value[col].isin(train[col])].copy()
    
    features_frames = []
    for feature in features:           
        feature_frame = df_value.reindex(columns=[col, feature])
        if feature in ["genres", "countries"]:
            feature_frame = feature_frame.explode(feature)
        feature_frame.columns = ["id", "value"]
        feature_frame["feature"] = feature
        features_frames.append(feature_frame)
    
    return pd.concat(features_frames)


def cross_val(model, cv, interactions_df, user_df, item_df, features, k_recos):
    """
        Cross validation for models
    """
    results = list()
    fold_iterator = cv.split(Interactions(interactions), collect_fold_stats=True)
    
    for i_fold, (train_ids, test_ids, fold_info) in enumerate(fold_iterator):
        print(f"\n==================== Fold {i_fold}")

        train = interactions.iloc[train_ids].copy()
        test = interactions.iloc[test_ids][Columns.UserItem].copy()
        
        train.drop(train.query("total_dur < 300").index, inplace=True)
        cold_users = set(test[Columns.User]) - set(train[Columns.User])
        test.drop(test[test[Columns.User].isin(cold_users)].index, inplace=True)
        
        user_features = create_features_dataset(Columns.User, train, user_df, features["users"])
        item_features = create_features_dataset(Columns.Item, train, item_df, features["items"])
        
        dataset = Dataset.construct(
            interactions_df=train,
            user_features_df=user_features,
            cat_user_features=features["users"],
            item_features_df=item_features,
            cat_item_features=features["items"]
        )
        
        user_test = test[Columns.User].unique() 
        catalog = train[Columns.Item].unique()
        
        model.fit(dataset)

        recos = model.recommend(
            users=user_test,
            dataset=dataset,
            k=k_recos,
            filter_viewed=True,
        )
        
        metric_values = calc_metrics(metrics, recos, test, train, catalog)
        results.append(metric_values) 
    
    df_result = pd.DataFrame(results)
    
    return df_result.mean(axis=0).tolist()


def objective_als(trial, config, cv, interactions_df, user_df, item_df, features, metrics):
    """
    Function for select hyperparameters for ALS model
    """
    MODEL = "ALS"
    
    n_factors = trial.suggest_categorical("n_factors", config["N_FACTORS"])
    
    is_fitting_features = trial.suggest_categorical(
        "is_fitting_features", 
        config["is_fitting_features"]
    )

    if f"{MODEL}_{n_factors}_{is_fitting_features}" in suggested_params:
        suggested_params[f"{MODEL}_{n_factors}_{is_fitting_features}"] += 1
        raise optuna.exceptions.TrialPruned()
    else:
        suggested_params[f"{MODEL}_{n_factors}_{is_fitting_features}"] = 1
    
    model = ImplicitALSWrapperModel(
        model=AlternatingLeastSquares(
            factors=n_factors, 
            random_state=config["RANDOM_STATE"], 
            num_threads=config["NUM_THREADS"],
        ), 
        fit_features_together=is_fitting_features
    )
    
    trial.set_user_attr("n_factors", n_factors)
    trial.set_user_attr("is_fitting_features", is_fitting_features)
    
    print(f"***RUN***\nmodel: {MODEL}\nN_FACTORS: {n_factors}\nis_fitting_features: {is_fitting_features}\n")
        
    return cross_val(model, cv, interactions_df, user_df, item_df, features, k_recos=config["K_RECOS"])


def objective_lfm(trial, config, cv, interactions_df, user_df, item_df, features, metrics):
    """
    Function for select hyperparameters for LightFM model
    """
    MODEL = "LightFM"
    
    n_factors = trial.suggest_categorical("n_factors", config["N_FACTORS"])
    losses = trial.suggest_categorical("losses", config["lightfm_losses"])
    user_alpha = trial.suggest_categorical("user_alpha", config["USER_ALPHA"])
    item_alpha = trial.suggest_categorical("item_alpha", config["ITEM_ALPHA"])
    
    if f"{MODEL}_{n_factors}_{losses}_{user_alpha}_{item_alpha}" in suggested_params:
        suggested_params[f"{MODEL}_{n_factors}_{losses}_{user_alpha}_{item_alpha}"] += 1
        raise optuna.exceptions.TrialPruned()
    else:
        suggested_params[f"{MODEL}_{n_factors}_{losses}_{user_alpha}_{item_alpha}"] = 1
    
    model = LightFMWrapperModel(
        model=LightFM(
            no_components=n_factors, 
            loss=losses, 
            random_state=config["RANDOM_STATE"], 
            learning_rate=config["LEARNING_RATE"], 
            user_alpha=user_alpha,
            item_alpha=item_alpha,
        ),
        epochs=config["N_EPOCHS"],
        num_threads=config["NUM_THREADS"],
    )
    
    trial.set_user_attr("n_factors", n_factors)
    trial.set_user_attr("losses", losses)
    trial.set_user_attr("user_alpha", user_alpha)
    trial.set_user_attr("losses", item_alpha)
    
    print(f"***RUN***model: {MODEL}\nN_FACTORS: {n_factors}\nloss: {losses}\nuser_alpha: {user_alpha}\nitem_alpha: {item_alpha}\n")

    return cross_val(model, cv, interactions_df, user_df, item_df, features, k_recos=config["K_RECOS"])

In [49]:
CONFIG_MODELS = {
    "K_RECOS": 10,
    "RANDOM_STATE": 42,
    "NUM_THREADS": 16,
    "N_FACTORS": (16, 32, ),
    "is_fitting_features": (False, True),  # ALS
    "N_EPOCHS": 10,  # Lightfm
    "USER_ALPHA": (0,),  # Lightfm 
    "ITEM_ALPHA": (0,),  # Lightfm
    "LEARNING_RATE": 0.05,  # Lightfm
    "lightfm_losses": ('bpr', 'warp', ),  # Lightfm
}

In [34]:
%%time

study_als = optuna.create_study(
    study_name="ALS",
    sampler=TPESampler(seed=42),
    directions=['maximize', 'maximize', 'maximize', 'maximize', 'maximize']
)
study_als.optimize(
    lambda trial: objective_als(
        trial=trial, 
        config=CONFIG_MODELS, 
        cv=cv, 
        interactions_df=interactions, 
        user_df=users, 
        item_df=items, 
        features=features, 
        metrics=metrics
    ),
    n_trials=int(
        len(CONFIG_MODELS["N_FACTORS"]) * 
        len(CONFIG_MODELS["is_fitting_features"]) * 
        2
    )
)

[I 2022-12-12 13:24:19,615] A new study created in memory with name: ALS


***RUN***
model: ALS
N_FACTORS: 32
is_fitting_features: False


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 13:29:55,593] Trial 0 finished with values: [0.033030692508210814, 0.15920183057692003, 0.0702827179628595, 5.17550196751981, 6.158320385214538e-05] and parameters: {'n_factors': 32, 'is_fitting_features': False}. 


***RUN***
model: ALS
N_FACTORS: 16
is_fitting_features: True


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 13:37:08,786] Trial 1 finished with values: [0.03202969841185801, 0.1608645253011474, 0.08259661188671252, 4.824391508634623, 2.8752851349528606e-05] and parameters: {'n_factors': 16, 'is_fitting_features': True}. 


***RUN***
model: ALS
N_FACTORS: 32
is_fitting_features: True


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 13:44:47,897] Trial 2 finished with values: [0.032152985846619395, 0.1615385029437911, 0.0831643603394286, 4.7999162605677075, 2.8024957212082277e-05] and parameters: {'n_factors': 32, 'is_fitting_features': True}. 
[I 2022-12-12 13:44:47,898] Trial 3 pruned. 
[I 2022-12-12 13:44:47,899] Trial 4 pruned. 
[I 2022-12-12 13:44:47,901] Trial 5 pruned. 
[I 2022-12-12 13:44:47,902] Trial 6 pruned. 


***RUN***
model: ALS
N_FACTORS: 16
is_fitting_features: False


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 13:49:55,795] Trial 7 finished with values: [0.033457762344784286, 0.1625549617602303, 0.07228444822316231, 4.860120719285789, 3.719975293722041e-05] and parameters: {'n_factors': 16, 'is_fitting_features': False}. 


CPU times: user 1h 1min 40s, sys: 15min 26s, total: 1h 17min 6s
Wall time: 25min 36s


In [38]:
metrics

{'prec@10': Precision(k=10),
 'recall@10': Recall(k=10),
 'MAP@10': MAP(k=10, divide_by_k=False),
 'novelty': MeanInvUserFreq(k=10),
 'serendipity': Serendipity(k=10)}

In [39]:
log_study_als = study_als.trials_dataframe()
log_study_als.rename(
    columns={
        "values_0": "prec@10", "values_1": "recall@10", "values_2": "MAP@10", "values_3": "novelty", 
        "values_4": "serendipity"
    }, 
    inplace=True
)
log_study_als

number   prec@10  recall@10    MAP@10   novelty  serendipity  \
0       0  0.033031   0.159202  0.070283  5.175502     0.000062   
1       1  0.032030   0.160865  0.082597  4.824392     0.000029   
2       2  0.032153   0.161539  0.083164  4.799916     0.000028   
3       3       NaN        NaN       NaN       NaN          NaN   
4       4       NaN        NaN       NaN       NaN          NaN   
5       5       NaN        NaN       NaN       NaN          NaN   
6       6       NaN        NaN       NaN       NaN          NaN   
7       7  0.033458   0.162555  0.072284  4.860121     0.000037   

              datetime_start          datetime_complete  \
0 2022-12-12 13:24:19.617119 2022-12-12 13:29:55.593706   
1 2022-12-12 13:29:55.594592 2022-12-12 13:37:08.785865   
2 2022-12-12 13:37:08.786772 2022-12-12 13:44:47.897142   
3 2022-12-12 13:44:47.898363 2022-12-12 13:44:47.898769   
4 2022-12-12 13:44:47.899255 2022-12-12 13:44:47.899665   
5 2022-12-12 13:44:47.900641 2022-12-12 13:44:47.900944   
6 2022-12-12 13:44:47.901530 2022-12-12 13:44:47.901959   
7 2022-12-12 13:44:47.902573 2022-12-12 13:49:55.795716   

                duration  params_is_fitting_features  params_n_factors  \
0 0 days 00:05:35.976587                       False                32   
1 0 days 00:07:13.191273                        True                16   
2 0 days 00:07:39.110370                        True                32   
3 0 days 00:00:00.000406                        True                16   
4 0 days 00:00:00.000410                       False                32   
5 0 days 00:00:00.000303                        True                16   
6 0 days 00:00:00.000429                        True                32   
7 0 days 00:05:07.893143                       False                16   

  user_attrs_is_fitting_features  user_attrs_n_factors     state  
0                          False                  32.0  COMPLETE  
1                           True                  16.0  COMPLETE  
2                           True                  32.0  COMPLETE  
3                            NaN                   NaN    PRUNED  
4                            NaN                   NaN    PRUNED  
5                            NaN                   NaN    PRUNED  
6                            NaN                   NaN    PRUNED  
7                          False                  16.0  COMPLETE

In [45]:
log_study_als.drop(log_study_als[log_study_als["state"] != "COMPLETE"].index, axis=0, inplace=True)
log_study_als

number   prec@10  recall@10    MAP@10   novelty  serendipity  \
0       0  0.033031   0.159202  0.070283  5.175502     0.000062   
1       1  0.032030   0.160865  0.082597  4.824392     0.000029   
2       2  0.032153   0.161539  0.083164  4.799916     0.000028   
7       7  0.033458   0.162555  0.072284  4.860121     0.000037   

              datetime_start          datetime_complete  \
0 2022-12-12 13:24:19.617119 2022-12-12 13:29:55.593706   
1 2022-12-12 13:29:55.594592 2022-12-12 13:37:08.785865   
2 2022-12-12 13:37:08.786772 2022-12-12 13:44:47.897142   
7 2022-12-12 13:44:47.902573 2022-12-12 13:49:55.795716   

                duration  params_is_fitting_features  params_n_factors  \
0 0 days 00:05:35.976587                       False                32   
1 0 days 00:07:13.191273                        True                16   
2 0 days 00:07:39.110370                        True                32   
7 0 days 00:05:07.893143                       False                16   

  user_attrs_is_fitting_features  user_attrs_n_factors     state  
0                          False                  32.0  COMPLETE  
1                           True                  16.0  COMPLETE  
2                           True                  32.0  COMPLETE  
7                          False                  16.0  COMPLETE

In [54]:
log_study_als["model"] = "ALS"
log_study_als.to_csv("../data/hw_4/log_study_als.csv", index=False)

In [51]:
%%time

study_lfm = optuna.create_study(
    study_name="LightFM",
    sampler=TPESampler(seed=42),
    directions=['maximize', 'maximize', 'maximize', 'maximize', 'maximize']
)
study_lfm.optimize(
    lambda trial: objective_lfm(
        trial=trial, 
        config=CONFIG_MODELS, 
        cv=cv, 
        interactions_df=interactions, 
        user_df=users, 
        item_df=items, 
        features=features, 
        metrics=metrics
    ),
    n_trials=int(
        len(CONFIG_MODELS["N_FACTORS"]) *
        len(CONFIG_MODELS["USER_ALPHA"]) * 
        len(CONFIG_MODELS["ITEM_ALPHA"]) *  
        len(CONFIG_MODELS["lightfm_losses"]) *
        2
    )
)

[I 2022-12-12 14:01:05,529] A new study created in memory with name: LightFM


***RUN***model: LightFM
N_FACTORS: 32
loss: bpr
user_alpha: 0
item_alpha: 0


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:07:31,301] Trial 0 finished with values: [0.0068209394096222925, 0.03101255683019741, 0.015230800287448587, 12.068082761810189, 0.00011798481324603613] and parameters: {'n_factors': 32, 'losses': 'bpr', 'user_alpha': 0, 'item_alpha': 0}. 


***RUN***model: LightFM
N_FACTORS: 16
loss: warp
user_alpha: 0
item_alpha: 0


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:10:10,527] Trial 1 finished with values: [0.03417229586777249, 0.1679148467818227, 0.08044566974378901, 5.1767501080041, 6.310733976519341e-05] and parameters: {'n_factors': 16, 'losses': 'warp', 'user_alpha': 0, 'item_alpha': 0}. 


***RUN***model: LightFM
N_FACTORS: 32
loss: warp
user_alpha: 0
item_alpha: 0


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:13:13,837] Trial 2 finished with values: [0.03387333769793409, 0.16619084263850772, 0.0790568310893865, 5.508329186408287, 9.017829915307326e-05] and parameters: {'n_factors': 32, 'losses': 'warp', 'user_alpha': 0, 'item_alpha': 0}. 
[I 2022-12-12 14:13:13,838] Trial 3 pruned. 
[I 2022-12-12 14:13:13,839] Trial 4 pruned. 
[I 2022-12-12 14:13:13,840] Trial 5 pruned. 
[I 2022-12-12 14:13:13,841] Trial 6 pruned. 


***RUN***model: LightFM
N_FACTORS: 16
loss: bpr
user_alpha: 0
item_alpha: 0


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:17:50,320] Trial 7 finished with values: [0.005295678169442308, 0.02416944628414099, 0.012153373034598803, 12.79936887733164, 8.33351124291126e-05] and parameters: {'n_factors': 16, 'losses': 'bpr', 'user_alpha': 0, 'item_alpha': 0}. 


CPU times: user 1h 32min 28s, sys: 2.67 s, total: 1h 32min 31s
Wall time: 16min 44s


In [55]:
log_study_lfm = study_lfm.trials_dataframe()
log_study_lfm.rename(
    columns={
        "values_0": "prec@10", "values_1": "recall@10", "values_2": "MAP@10", "values_3": "novelty", 
        "values_4": "serendipity"
    }, 
    inplace=True
)
log_study_lfm.drop(log_study_lfm[log_study_lfm["state"] != "COMPLETE"].index, axis=0, inplace=True)
log_study_lfm["model"] = "LighFM"

In [56]:
log_study_lfm

number   prec@10  recall@10    MAP@10    novelty  serendipity  \
0       0  0.006821   0.031013  0.015231  12.068083     0.000118   
1       1  0.034172   0.167915  0.080446   5.176750     0.000063   
2       2  0.033873   0.166191  0.079057   5.508329     0.000090   
7       7  0.005296   0.024169  0.012153  12.799369     0.000083   

              datetime_start          datetime_complete  \
0 2022-12-12 14:01:05.530755 2022-12-12 14:07:31.301238   
1 2022-12-12 14:07:31.302396 2022-12-12 14:10:10.527654   
2 2022-12-12 14:10:10.528535 2022-12-12 14:13:13.837143   
7 2022-12-12 14:13:13.842459 2022-12-12 14:17:50.319895   

                duration  params_item_alpha params_losses  params_n_factors  \
0 0 days 00:06:25.770483                  0           bpr                32   
1 0 days 00:02:39.225258                  0          warp                16   
2 0 days 00:03:03.308608                  0          warp                32   
7 0 days 00:04:36.477436                  0           bpr                16   

   params_user_alpha  user_attrs_losses  user_attrs_n_factors  \
0                  0                0.0                  32.0   
1                  0                0.0                  16.0   
2                  0                0.0                  32.0   
7                  0                0.0                  16.0   

   user_attrs_user_alpha     state   model  
0                    0.0  COMPLETE  LighFM  
1                    0.0  COMPLETE  LighFM  
2                    0.0  COMPLETE  LighFM  
7                    0.0  COMPLETE  LighFM

In [57]:
log_study_lfm.to_csv("../data/hw_4/log_study_lfm.csv", index=False)

**Лучшие результаты обучения показывает модель lighFM с лоссом warp** подберем для моделей с данным лоссом оптимальное user_alpha и item_alpha

In [58]:
CONFIG_MODELS_LFM = {
    "K_RECOS": 10,
    "RANDOM_STATE": 42,
    "NUM_THREADS": 16,
    "N_FACTORS": (16, 32, ),
    "N_EPOCHS": 10,  # Lightfm
    "USER_ALPHA": (0, 0.2),  # Lightfm 
    "ITEM_ALPHA": (0, 0.2),  # Lightfm
    "LEARNING_RATE": 0.05,  # Lightfm
    "lightfm_losses": ('warp', ),  # Lightfm
}

In [59]:
%%time

study_lfm_alpha = optuna.create_study(
    study_name="LightFM_alpha",
    sampler=TPESampler(seed=42),
    directions=['maximize', 'maximize', 'maximize', 'maximize', 'maximize']
)
study_lfm_alpha.optimize(
    lambda trial: objective_lfm(
        trial=trial, 
        config=CONFIG_MODELS_LFM, 
        cv=cv, 
        interactions_df=interactions, 
        user_df=users, 
        item_df=items, 
        features=features, 
        metrics=metrics
    ),
    n_trials=int(
        len(CONFIG_MODELS["N_FACTORS"]) *
        len(CONFIG_MODELS["USER_ALPHA"]) * 
        len(CONFIG_MODELS["ITEM_ALPHA"]) *
        2
    )
)

[I 2022-12-12 14:25:21,276] A new study created in memory with name: LightFM_alpha
[I 2022-12-12 14:25:21,279] Trial 0 pruned. 


***RUN***model: LightFM
N_FACTORS: 32
loss: warp
user_alpha: 0.2
item_alpha: 0.2


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:29:36,116] Trial 1 finished with values: [0.034278189254484864, 0.17449544816064208, 0.08483012091477123, 3.766206685818005, 2.354521448242052e-06] and parameters: {'n_factors': 32, 'losses': 'warp', 'user_alpha': 0.2, 'item_alpha': 0.2}. 


***RUN***model: LightFM
N_FACTORS: 16
loss: warp
user_alpha: 0.2
item_alpha: 0.2


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:32:43,095] Trial 2 finished with values: [0.0344082365349989, 0.175267733724092, 0.08524555380380662, 3.7785924768795187, 2.3107020986006132e-06] and parameters: {'n_factors': 16, 'losses': 'warp', 'user_alpha': 0.2, 'item_alpha': 0.2}. 


***RUN***model: LightFM
N_FACTORS: 16
loss: warp
user_alpha: 0
item_alpha: 0.2


==================== Fold 0

==================== Fold 1

==================== Fold 2


[I 2022-12-12 14:37:16,325] Trial 3 finished with values: [8.121216487323505e-05, 0.00023670865324451867, 6.592145604060793e-05, 16.300173367831395, 4.115795688696286e-06] and parameters: {'n_factors': 16, 'losses': 'warp', 'user_alpha': 0, 'item_alpha': 0.2}. 


CPU times: user 1h 1min 35s, sys: 2.72 s, total: 1h 1min 38s
Wall time: 11min 55s


In [60]:
log_study_lfm_alpha = study_lfm_alpha.trials_dataframe()
log_study_lfm_alpha.rename(
    columns={
        "values_0": "prec@10", "values_1": "recall@10", "values_2": "MAP@10", "values_3": "novelty", 
        "values_4": "serendipity"
    }, 
    inplace=True
)
log_study_lfm_alpha.drop(
    log_study_lfm_alpha[log_study_lfm_alpha["state"] != "COMPLETE"].index, axis=0, inplace=True
)
log_study_lfm_alpha["model"] = "LighFM"
log_study_lfm_alpha

number   prec@10  recall@10    MAP@10    novelty  serendipity  \
1       1  0.034278   0.174495  0.084830   3.766207     0.000002   
2       2  0.034408   0.175268  0.085246   3.778592     0.000002   
3       3  0.000081   0.000237  0.000066  16.300173     0.000004   

              datetime_start          datetime_complete  \
1 2022-12-12 14:25:21.279916 2022-12-12 14:29:36.116148   
2 2022-12-12 14:29:36.116969 2022-12-12 14:32:43.095291   
3 2022-12-12 14:32:43.096137 2022-12-12 14:37:16.325048   

                duration  params_item_alpha params_losses  params_n_factors  \
1 0 days 00:04:14.836232                0.2          warp                32   
2 0 days 00:03:06.978322                0.2          warp                16   
3 0 days 00:04:33.228911                0.2          warp                16   

   params_user_alpha  user_attrs_losses  user_attrs_n_factors  \
1                0.2                0.2                  32.0   
2                0.2                0.2                  16.0   
3                0.0                0.2                  16.0   

   user_attrs_user_alpha     state   model  
1                    0.2  COMPLETE  LighFM  
2                    0.2  COMPLETE  LighFM  
3                    0.0  COMPLETE  LighFM

In [61]:
log_study_lfm = pd.concat([log_study_lfm, log_study_lfm_alpha], ignore_index=True)
log_study_lfm

number   prec@10  recall@10    MAP@10    novelty  serendipity  \
0       0  0.006821   0.031013  0.015231  12.068083     0.000118   
1       1  0.034172   0.167915  0.080446   5.176750     0.000063   
2       2  0.033873   0.166191  0.079057   5.508329     0.000090   
3       7  0.005296   0.024169  0.012153  12.799369     0.000083   
4       1  0.034278   0.174495  0.084830   3.766207     0.000002   
5       2  0.034408   0.175268  0.085246   3.778592     0.000002   
6       3  0.000081   0.000237  0.000066  16.300173     0.000004   

              datetime_start          datetime_complete  \
0 2022-12-12 14:01:05.530755 2022-12-12 14:07:31.301238   
1 2022-12-12 14:07:31.302396 2022-12-12 14:10:10.527654   
2 2022-12-12 14:10:10.528535 2022-12-12 14:13:13.837143   
3 2022-12-12 14:13:13.842459 2022-12-12 14:17:50.319895   
4 2022-12-12 14:25:21.279916 2022-12-12 14:29:36.116148   
5 2022-12-12 14:29:36.116969 2022-12-12 14:32:43.095291   
6 2022-12-12 14:32:43.096137 2022-12-12 14:37:16.325048   

                duration  params_item_alpha params_losses  params_n_factors  \
0 0 days 00:06:25.770483                0.0           bpr                32   
1 0 days 00:02:39.225258                0.0          warp                16   
2 0 days 00:03:03.308608                0.0          warp                32   
3 0 days 00:04:36.477436                0.0           bpr                16   
4 0 days 00:04:14.836232                0.2          warp                32   
5 0 days 00:03:06.978322                0.2          warp                16   
6 0 days 00:04:33.228911                0.2          warp                16   

   params_user_alpha  user_attrs_losses  user_attrs_n_factors  \
0                0.0                0.0                  32.0   
1                0.0                0.0                  16.0   
2                0.0                0.0                  32.0   
3                0.0                0.0                  16.0   
4                0.2                0.2                  32.0   
5                0.2                0.2                  16.0   
6                0.0                0.2                  16.0   

   user_attrs_user_alpha     state   model  
0                    0.0  COMPLETE  LighFM  
1                    0.0  COMPLETE  LighFM  
2                    0.0  COMPLETE  LighFM  
3                    0.0  COMPLETE  LighFM  
4                    0.2  COMPLETE  LighFM  
5                    0.2  COMPLETE  LighFM  
6                    0.0  COMPLETE  LighFM

In [62]:
log_study_lfm.to_csv("../data/hw_4/log_study_lfm.csv", index=False)

# Анализ метрик

In [63]:
log_study_als = pd.read_csv("../data/hw_4/log_study_als.csv")
log_study_lfm = pd.read_csv("../data/hw_4/log_study_lfm.csv")

In [64]:
log_study_als

number   prec@10  recall@10    MAP@10   novelty  serendipity  \
0       0  0.033031   0.159202  0.070283  5.175502     0.000062   
1       1  0.032030   0.160865  0.082597  4.824392     0.000029   
2       2  0.032153   0.161539  0.083164  4.799916     0.000028   
3       7  0.033458   0.162555  0.072284  4.860121     0.000037   

               datetime_start           datetime_complete  \
0  2022-12-12 13:24:19.617119  2022-12-12 13:29:55.593706   
1  2022-12-12 13:29:55.594592  2022-12-12 13:37:08.785865   
2  2022-12-12 13:37:08.786772  2022-12-12 13:44:47.897142   
3  2022-12-12 13:44:47.902573  2022-12-12 13:49:55.795716   

                 duration  params_is_fitting_features  params_n_factors  \
0  0 days 00:05:35.976587                       False                32   
1  0 days 00:07:13.191273                        True                16   
2  0 days 00:07:39.110370                        True                32   
3  0 days 00:05:07.893143                       False                16   

   user_attrs_is_fitting_features  user_attrs_n_factors     state model  
0                           False                  32.0  COMPLETE   ALS  
1                            True                  16.0  COMPLETE   ALS  
2                            True                  32.0  COMPLETE   ALS  
3                           False                  16.0  COMPLETE   ALS

In [65]:
log_study_lfm

number   prec@10  recall@10    MAP@10    novelty  serendipity  \
0       0  0.006821   0.031013  0.015231  12.068083     0.000118   
1       1  0.034172   0.167915  0.080446   5.176750     0.000063   
2       2  0.033873   0.166191  0.079057   5.508329     0.000090   
3       7  0.005296   0.024169  0.012153  12.799369     0.000083   
4       1  0.034278   0.174495  0.084830   3.766207     0.000002   
5       2  0.034408   0.175268  0.085246   3.778592     0.000002   
6       3  0.000081   0.000237  0.000066  16.300173     0.000004   

               datetime_start           datetime_complete  \
0  2022-12-12 14:01:05.530755  2022-12-12 14:07:31.301238   
1  2022-12-12 14:07:31.302396  2022-12-12 14:10:10.527654   
2  2022-12-12 14:10:10.528535  2022-12-12 14:13:13.837143   
3  2022-12-12 14:13:13.842459  2022-12-12 14:17:50.319895   
4  2022-12-12 14:25:21.279916  2022-12-12 14:29:36.116148   
5  2022-12-12 14:29:36.116969  2022-12-12 14:32:43.095291   
6  2022-12-12 14:32:43.096137  2022-12-12 14:37:16.325048   

                 duration  params_item_alpha params_losses  params_n_factors  \
0  0 days 00:06:25.770483                0.0           bpr                32   
1  0 days 00:02:39.225258                0.0          warp                16   
2  0 days 00:03:03.308608                0.0          warp                32   
3  0 days 00:04:36.477436                0.0           bpr                16   
4  0 days 00:04:14.836232                0.2          warp                32   
5  0 days 00:03:06.978322                0.2          warp                16   
6  0 days 00:04:33.228911                0.2          warp                16   

   params_user_alpha  user_attrs_losses  user_attrs_n_factors  \
0                0.0                0.0                  32.0   
1                0.0                0.0                  16.0   
2                0.0                0.0                  32.0   
3                0.0                0.0                  16.0   
4                0.2                0.2                  32.0   
5                0.2                0.2                  16.0   
6                0.0                0.2                  16.0   

   user_attrs_user_alpha     state   model  
0                    0.0  COMPLETE  LighFM  
1                    0.0  COMPLETE  LighFM  
2                    0.0  COMPLETE  LighFM  
3                    0.0  COMPLETE  LighFM  
4                    0.2  COMPLETE  LighFM  
5                    0.2  COMPLETE  LighFM  
6                    0.0  COMPLETE  LighFM

In [68]:
pd.concat([log_study_als, log_study_lfm], ignore_index=True)

number   prec@10  recall@10    MAP@10    novelty  serendipity  \
0        0  0.033031   0.159202  0.070283   5.175502     0.000062   
1        1  0.032030   0.160865  0.082597   4.824392     0.000029   
2        2  0.032153   0.161539  0.083164   4.799916     0.000028   
3        7  0.033458   0.162555  0.072284   4.860121     0.000037   
4        0  0.006821   0.031013  0.015231  12.068083     0.000118   
5        1  0.034172   0.167915  0.080446   5.176750     0.000063   
6        2  0.033873   0.166191  0.079057   5.508329     0.000090   
7        7  0.005296   0.024169  0.012153  12.799369     0.000083   
8        1  0.034278   0.174495  0.084830   3.766207     0.000002   
9        2  0.034408   0.175268  0.085246   3.778592     0.000002   
10       3  0.000081   0.000237  0.000066  16.300173     0.000004   

                datetime_start           datetime_complete  \
0   2022-12-12 13:24:19.617119  2022-12-12 13:29:55.593706   
1   2022-12-12 13:29:55.594592  2022-12-12 13:37:08.785865   
2   2022-12-12 13:37:08.786772  2022-12-12 13:44:47.897142   
3   2022-12-12 13:44:47.902573  2022-12-12 13:49:55.795716   
4   2022-12-12 14:01:05.530755  2022-12-12 14:07:31.301238   
5   2022-12-12 14:07:31.302396  2022-12-12 14:10:10.527654   
6   2022-12-12 14:10:10.528535  2022-12-12 14:13:13.837143   
7   2022-12-12 14:13:13.842459  2022-12-12 14:17:50.319895   
8   2022-12-12 14:25:21.279916  2022-12-12 14:29:36.116148   
9   2022-12-12 14:29:36.116969  2022-12-12 14:32:43.095291   
10  2022-12-12 14:32:43.096137  2022-12-12 14:37:16.325048   

                  duration params_is_fitting_features  params_n_factors  \
0   0 days 00:05:35.976587                      False                32   
1   0 days 00:07:13.191273                       True                16   
2   0 days 00:07:39.110370                       True                32   
3   0 days 00:05:07.893143                      False                16   
4   0 days 00:06:25.770483                        NaN                32   
5   0 days 00:02:39.225258                        NaN                16   
6   0 days 00:03:03.308608                        NaN                32   
7   0 days 00:04:36.477436                        NaN                16   
8   0 days 00:04:14.836232                        NaN                32   
9   0 days 00:03:06.978322                        NaN                16   
10  0 days 00:04:33.228911                        NaN                16   

   user_attrs_is_fitting_features  user_attrs_n_factors     state   model  \
0                           False                  32.0  COMPLETE     ALS   
1                            True                  16.0  COMPLETE     ALS   
2                            True                  32.0  COMPLETE     ALS   
3                           False                  16.0  COMPLETE     ALS   
4                             NaN                  32.0  COMPLETE  LighFM   
5                             NaN                  16.0  COMPLETE  LighFM   
6                             NaN                  32.0  COMPLETE  LighFM   
7                             NaN                  16.0  COMPLETE  LighFM   
8                             NaN                  32.0  COMPLETE  LighFM   
9                             NaN                  16.0  COMPLETE  LighFM   
10                            NaN                  16.0  COMPLETE  LighFM   

    params_item_alpha params_losses  params_user_alpha  user_attrs_losses  \
0                 NaN           NaN                NaN                NaN   
1                 NaN           NaN                NaN                NaN   
2                 NaN           NaN                NaN                NaN   
3                 NaN           NaN                NaN                NaN   
4                 0.0           bpr                0.0                0.0   
5                 0.0          warp                0.0                0.0   
6                 0.0          warp                0.0           

По метрике **MAP@10** выберем по одной лучшей модели из ALS и LighFM:

**ALS MAP@10 = 0.083164** 
 - params_is_fitting_features = True
 - params_n_factors = 32
 
**LighFM MAP@10 = 0.085246** 
 - params_n_factors = 16
 - params_losses = warp
 - params_item_alpha = 0.2
 - params_user_alpha = 0.2

# Обучение лучших моделей с валидации, добавление 3-ёх аватаров, метод приближенного поиска соседей при помощи nmslib

## Формирование векторов фичей юзеров и айтемов

In [74]:
interactions.drop(interactions.query("total_dur < 300").index, inplace=True)

**Добавим 3 аватаров**:
1. Возраст от 45-54, доход 60-90, пол мужской, смотрим русские сериалы боевики (id 5555555)
2. Возраст от 18-24, доход 20-40, пол мужской, смотрит фильмы жанра комедия страна любая кроме России (id 6666666)
3. Возраст от 35-44, доход 40-60, пол женский, смотрим и фильмы и сериалы жанра мелодрамы неважна какая страна (id 7777777)

Добавили именно таких юзеров, т.к. можно между ними провести грань, например, предлагаем, что для первого аватара не попадёт в выдачу сериалы с жанром не боевик и будут только русские сериала, возможно в выдачу могут попасть русские фильмы боевики. Для второго пользователя должны рекомендоваться комедии не из России. Для Третьего только мелодрамы

In [142]:
first_cond = items[items["countries"].apply(lambda countries: "россия" in countries)]["item_id"].to_numpy()
second_cond = items[items["genres"].apply(lambda genres: "боевики" in genres)]["item_id"].to_numpy()
first_avatar_items = items[
    (
        (items["content_type"] == "series") & 
        (items["item_id"].isin(first_cond)) &
        (items["item_id"].isin(second_cond))
    )
].reset_index(drop=True)
display_df(first_avatar_items)

item_id content_type                          title  \
0      8434       series               Застывшие депеши   
1      2074       series                      Посредник   
2       415       series        Приступить к ликвидации   
3      3530       series                Господа офицеры   
4     10839       series                      Рокировка   
74    15443       series                        Морпехи   
75     8632       series           Бой с тенью (сериал)   
76     9591       series                  Паршивые овцы   
77    14389       series  Последний бой майора Пугачёва   
78     2367       series                        Надежда   

                        title_orig  release_year  \
0                              NaN        2010.0   
1                        Posrednik        1990.0   
2          Pristupit k likvidatsii        1983.0   
3                              NaN        2004.0   
4                              NaN        2004.0   
74                             NaN        2011.0   
75                     Boy s tenyu        2005.0   
76                             NaN        2010.0   
77  Posledniy boy mayora Pugachova        2005.0   
78                             NaN        2020.0   

                                genres countries  for_kids  age_rating  \
0        [боевики, русские, детективы]  [россия]       NaN        16.0   
1                [боевики, фантастика]  [россия]       NaN        16.0   
2                  [боевики, криминал]  [россия]       NaN         6.0   
3                   [боевики, русские]  [россия]       NaN        16.0   
4        [боевики, русские, мелодрамы]  [россия]       NaN        12.0   
74  [драмы, боевики, русские, военные]  [россия]       NaN        16.0   
75   [боевики, драмы, спорт, криминал]  [россия]       NaN        16.0   
76         [боевики, русские, военные]  [россия]       NaN        16.0   
77           [боевики, драмы, военные]  [россия]       NaN        16.0   
78                    [драмы, боевики]  [россия]       0.0        18.0   

   studios                                          directors  \
0      NaN                                     Армен Арутюнян   
1      NaN                                          В.Потапов   
2      NaN                                       Б. Григорьев   
3      NaN                                     Андрей Кравчук   
4      NaN  Владимир Кононенко, Милиана Черкасова, Тимофей...   
74     NaN                         Дмитрий Матов, Мурад Алиев   
75     NaN                                    Алексей Сидоров   
76     NaN                                     Сергей Чекалов   
77     NaN                                  Владимир Фатьянов   
78     NaN                                     Елена Хазанова   

                                               actors  \
0   Александр Орловский, Анна Литкенс, Артем Семак...   
1   Олеся Судзиловская, Инара Слуцка, Валерий Стор...   
2   Олег Стриженов, Михаил Жигалов, Василий Ланово...   
3   Александр Иванов, Аркадий Коваль, Артем Алексе...   
4   Александр Баранов, Анатолий Чижиков, Борис Мир...   
74  Александр Зельский, Александр Лобанов, Алексан...   
75  Денис Никифоров, Елена Панова, Иван Макаревич,...   
76  Алексей Дмитриев, Алексей Лонгин, Андрей Фроло...   
77  Игорь Лифанов, Борис Токарев, Лев Дуров, Игорь...   
78  Виктория Исакова, Александр Кузьмин, Алексей М...   

                                          description  \
0   Криминальная сага режиссера Армена Арутюняна с...   
1   Нависший над землей загадочный шар становится ...   
2   Весна 1945 года. Близится к концу Великая Отеч...   
3   Захватывающий многосерийный боевик, рассказыва...   
4   Успешный бизнесмен Игорь Березин в одночасье л...   
74  «Черные береты», «черные дьяволы», «черная сме...   
75  Артем Колчин – профессиональный боксер и главн...   
76  Сложившийся в работе над телевизионным сериало...   
77  Действие фильма начинается на территории Польш...   
78  Оригинальный киносериал от создателей «Бывших»..

In [143]:
first_cond = items[items["countries"].apply(lambda countries: "россия" not in countries)]["item_id"].to_numpy()
second_cond = items[items["genres"].apply(lambda genres: "комедии" in genres)]["item_id"].to_numpy()
second_avatar_items = items[
    (
        (items["content_type"] == "film") & 
        (items["item_id"].isin(first_cond)) &
        (items["item_id"].isin(second_cond))
    )
].reset_index(drop=True)
display_df(second_avatar_items)

item_id content_type                              title  \
0        2508         film                        Голые перцы   
1       10716         film                   Тактическая сила   
2        6181         film  Первая встреча, последняя встреча   
3       15076         film                           Бладфест   
4       13109         film             Новый парень моей мамы   
2712     8233         film      4.3.2.1 (с тифлокомментарием)   
2713     3563         film                   Утро без отметок   
2714    13074         film             Маленькая мисс Дулиттл   
2715    10372         film                      Правила секса   
2716     3364         film                      Восьмой класс   

                                       title_orig  release_year  \
0                                    Search Party        2014.0   
1                                  Tactical Force        2011.0   
2                                             NaN        1987.0   
3                                      Blood Fest        2018.0   
4                          My Mom's New Boyfriend        2007.0   
2712             4.3.2.1 (with Audio Description)        2010.0   
2713                             Utro bez otmetok        1983.0   
2714  Liliane Susewind - Ein tierisches Abenteuer        2018.0   
2715                     RULES OF ATTRACTION, THE        2002.0   
2716                                 Eighth Grade        2018.0   

                                                 genres         countries  \
0                    [зарубежные, приключения, комедии]             [сша]   
1     [криминал, зарубежные, триллеры, боевики, коме...          [канада]   
2                           [драмы, советские, комедии]            [ссср]   
3                          [зарубежные, ужасы, комедии]             [сша]   
4            [мелодрамы, зарубежные, криминал, комедии]        [германия]   
2712                    [зарубежные, триллеры, комедии]  [великобритания]   
2713                                [семейное, комедии]            [ссср]   
2714             [фэнтези, семейное, криминал, комедии]        [германия]   
2715                        [драмы, мелодрамы, комедии]   [германия, сша]   
2716                                   [драмы, комедии]             [сша]   

      for_kids  age_rating   studios                directors  \
0          NaN        16.0       NaN           Скот Армстронг   
1          NaN        16.0       NaN         Адам П. Калтраро   
2          NaN        16.0  Ленфильм        Виталий Мельников   
3          NaN        18.0       NaN             Оуэн Эгертон   
4          NaN        12.0       NaN             Джордж Галло   
2712       NaN        16.0       NaN  Ноэль Кларк, Марк Дэвис   
2713       NaN         6.0       NaN        Владимир Мартынов   
2714       NaN         6.0       NaN          Йоахим Масаннек   
2715       NaN        18.0       NaN             Роджер Эвери   
2716       NaN        16.0       NaN                Бо Бёрнэм   

                                                 actors  \
0     Адам Палли, Брайан Хаски, Дж.Б. Смув, Джейсон ...   
1     Адриан Холмс, Даррен Шалави, Джерри Вассерман,...   
2     Борис Плотников, Гражина Шаполовска, Иннокенти...   
3     Барбара Дункельман, Джейкоб Баталон, Крис Дабе...   
4     Антонио Бандерас, Джон Вальдетеро, Кит Дэвид, ...   
2712  Адам Дикон, Офелия Ловибонд, Тэмзин Эджертон, ...   
2713  Кирилл Головко-Серский, Мария Вартикова, Павел...   
2714  Малу Ляйхер, Пери Баумайстер, Том Бек, Айлин Т...   
2715  Джеймс Ван Дер Бик, Шаннин Соссамон, Джессика ...   
2716  Элси Фишер, Джош Хэмилтон, Эмили Робинсон, Дже...   

                                            description  \
0     Уморительная современная комедия на популярную...   
1     Профессиональный рестлер Стив Остин («Все или ...   
2     Инициатива наказуема, особенно в Петрограде вр...   
3     Фанат хорорров Дакс вместе с друзьями отправля...   
4     «Новый парень моей мамы» – американо-

In [144]:
first_cond = items[items["genres"].apply(lambda genres: "мелодрамы" in genres)]["item_id"].to_numpy()
third_avatar_items = items[
    (
        (items["item_id"].isin(first_cond))
    )
].reset_index(drop=True)
display_df(third_avatar_items)

item_id content_type                            title  \
0       10711         film                   Поговори с ней   
1        7868         film                           45 лет   
2       16268         film             Все решает мгновение   
3        8604         film                   Третья попытка   
4        3526         film                    Код «Красный»   
2773     5473         film  Чего ждать, когда ждёшь ребенка   
2774    10372         film                    Правила секса   
2775     1922         film                         Цзюй Доу   
2776     4430       series                      Наши соседи   
2777    15610       series                       Цена любви   

                                title_orig  release_year  \
0                           Hable con ella        2002.0   
1                                 45 Years        2015.0   
2                                      NaN        1978.0   
3                                      NaN        2013.0   
4                                 Red Joan        2018.0   
2773  WHAT TO EXPECT WHEN YOU'RE EXPECTING        2012.0   
2774              RULES OF ATTRACTION, THE        2002.0   
2775                                Ju Dou        1990.0   
2776                                   NaN        2010.0   
2777                           Tsena Lubvi        2015.0   

                                                 genres         countries  \
0             [драмы, зарубежные, детективы, мелодрамы]         [испания]   
1                        [драмы, зарубежные, мелодрамы]  [великобритания]   
2                  [драмы, спорт, советские, мелодрамы]            [ссср]   
3                                  [русские, мелодрамы]          [россия]   
4     [биография, экранизации, драмы, зарубежные, ме...  [великобритания]   
2773                                 [драмы, мелодрамы]             [сша]   
2774                        [драмы, мелодрамы, комедии]   [германия, сша]   
2775                                 [драмы, мелодрамы]           [китай]   
2776                               [русские, мелодрамы]          [россия]   
2777                                        [мелодрамы]          [россия]   

      for_kids  age_rating   studios                           directors  \
0          NaN        16.0       NaN                    Педро Альмодовар   
1          NaN        16.0       NaN                           Эндрю Хэй   
2          NaN        12.0  Ленфильм                    Виктор Садовский   
3          NaN        12.0       NaN                      Игорь Мужжухин   
4          NaN        18.0       NaN                         Тревор Нанн   
2773       NaN        16.0       NaN                          Кирк Джонс   
2774       NaN        18.0       NaN                        Роджер Эвери   
2775       NaN        12.0       NaN                           Чжан Имоу   
2776       NaN        16.0       NaN  Дмитрий Гольдман, Казбек Меретуков   
2777       NaN        16.0       NaN                             А. Хван   

                                                 actors  \
0     Адольфо Фернандес, Ана Фернандес, Дарио Гранди...   
1     Александра Риддлстон-Барретт, Джеральдин Джейм...   
2     Александр Абдулов, Александр Демьяненко, Алекс...   
3     Александр Асташенок, Александр Пашков, Андрей ...   
4     Бен Майлз, Джуди Денч, Лоуренс Спэллман, Софи ...   
2773  Кэмерон Диаз, Дженнифер Лопес, Элизабет Бэнкс,...   
2774  Джеймс Ван Дер Бик, Шаннин Соссамон, Джессика ...   
2775  Гун Ли, Ли Баотянь, Ли Вэй, Чжан И, Цзи-ан Чжэ...   
2776  Александр Зельский, Александр Шаврин, Амбарцум...   
2777  Анна Невская, Юрий Батурин, Анастасия Матвеева...   

                                            description  \
0     Мелодрама легендарного Педро Альмодовара «Пого...   
1     Шарлотта Рэмплинг, Том Кортни, Джеральдин Джей...   
2     Расчетливая чаровница из советского кинохита «...   
3     Екатерина Рябова, Александр Асташенок и Алекса...   
4     Тихая английска

In [124]:
7777777 in users[Columns.User]

False

In [125]:
6666666 in users[Columns.User]

False

In [126]:
5555555 in users[Columns.User]

False

In [127]:
users

user_id         age        income      sex  kids_flg
0        973171   age_25_34  income_60_90        М         1
1        962099   age_18_24  income_20_40        М         0
2       1047345   age_45_54  income_40_60        Ж         0
3        721985   age_45_54  income_20_40        Ж         0
4        704055   age_35_44  income_60_90        Ж         0
...         ...         ...           ...      ...       ...
840192   339025  age_65_inf   income_0_20        Ж         0
840193   983617   age_18_24  income_20_40        Ж         1
840194   251008     Unknown       Unknown  Unknown         0
840195   590706     Unknown       Unknown        Ж         0
840196   166555  age_65_inf  income_20_40        Ж         0

[840197 rows x 5 columns]

In [77]:
items_features = create_features_dataset(Columns.Item, interactions, items, features["items"])
display_df(items_features)

id     value       feature
0      10711      film  content_type
1       2508      film  content_type
2      10716      film  content_type
3       7868      film  content_type
4      16268      film  content_type
15958   6443  германия     countries
15959   2367    россия     countries
15960  10632    россия     countries
15961   4538    россия     countries
15962   3206    россия     countries

In [134]:
users_with_avatar = users.append(
    pd.DataFrame(
        {
            "user_id": [5555555, 6666666, 7777777],
            "age": ["age_45_54", "age_18_24", "age_35_44"],
            "income": ["income_60_90", "income_20_40", "income_40_60"],
            "sex": ["М", "М", "Ж"],
            "kids_flg": [0, 0, 0]    
        },
    ),
    ignore_index=True
)

users_with_avatar.tail(5)

user_id         age        income sex  kids_flg
840195   590706     Unknown       Unknown   Ж         0
840196   166555  age_65_inf  income_20_40   Ж         0
840197  5555555   age_45_54  income_60_90   М         0
840198  6666666   age_18_24  income_20_40   М         0
840199  7777777   age_35_44  income_40_60   Ж         0

In [135]:
display_df(interactions)

user_id  item_id last_watch_dt  total_dur  watched_pct  weight
0         176549     9506    2021-05-11       4250         72.0     5.0
1         699317     1659    2021-05-29       8317        100.0     6.0
3         864613     7638    2021-07-05      14483        100.0     6.0
4         964868     9506    2021-04-30       6725        100.0     6.0
5        1032142     6686    2021-05-13      11286        100.0     6.0
5476245   786732     4880    2021-05-12        753          0.0     0.0
5476247   546862     9673    2021-04-13       2308         49.0     4.0
5476248   697262    15297    2021-08-20      18307         63.0     4.0
5476249   384202    16197    2021-04-19       6203        100.0     6.0
5476250   319709     4436    2021-08-15       3921         45.0     4.0

In [140]:
interactions["user_id"].value_counts().describe()

count    809577.000000
mean          5.184158
std           9.285684
min           1.000000
25%           1.000000
50%           2.000000
75%           5.000000
max         461.000000
Name: user_id, dtype: float64

Отберем по **20 item** для каждого нового аватара

In [145]:
first_avatar_items = first_avatar_items.sample(20, random_state=17)["item_id"].to_numpy()
second_avatar_items = second_avatar_items.sample(20, random_state=17)["item_id"].to_numpy()
third_avatar_items = third_avatar_items.sample(20, random_state=17)["item_id"].to_numpy()

In [147]:
interactions['total_dur'].describe()

count    4.196975e+06
mean     1.079820e+04
std      5.523617e+04
min      3.000000e+02
25%      1.646000e+03
50%      5.353000e+03
75%      8.245000e+03
max      8.041167e+07
Name: total_dur, dtype: float64

In [162]:
total_dur = [np.random.randint(500, 10000) for _ in range(60)]

In [161]:
watched_pct, weight = [], []
for _ in range(60):
    random_value = interactions.sample(1)[["watched_pct", "weight"]].values[0]
    watched_pct.append(random_value[0])
    weight.append(random_value[1])

In [173]:
datetime = [
    interactions[interactions[Columns.Datetime] > "2021-05-01"].sample(1)[Columns.Datetime].iloc[0]
    for _ in range(60)
]

In [186]:
interactions_new_avatar = pd.DataFrame(
    {
        Columns.User: [5555555, 6666666, 7777777],
        Columns.Item: [list(first_avatar_items), list(second_avatar_items), list(third_avatar_items)],
        Columns.Datetime: [datetime[:20], datetime[20:40], datetime[40:]],
        "total_dur": [total_dur[:20], total_dur[20:40], total_dur[40:]],
        "watched_pct": [watched_pct[:20], watched_pct[20:40], watched_pct[40:]],
        Columns.Weight: [weight[:20], weight[20:40], weight[40:]],
    }
).explode(
    [Columns.Item, Columns.Datetime, "total_dur", "watched_pct", Columns.Weight]
).reset_index(drop=True)

interactions_new_avatar

user_id item_id last_watch_dt total_dur watched_pct weight
0   5555555     879    2021-06-21      5230        94.0    5.0
1   5555555    2280    2021-08-21      5476        22.0    3.0
2   5555555    3178    2021-07-04      4843        16.0    3.0
3   5555555   15813    2021-07-24      7414       100.0    6.0
4   5555555    3471    2021-06-27      8583       100.0    6.0
5   5555555    2113    2021-07-20      3288        63.0    4.0
6   5555555    3818    2021-08-20      6726        80.0    5.0
7   5555555    9550    2021-06-14      2662        16.0    3.0
8   5555555    3209    2021-08-16      5060       100.0    6.0
9   5555555    1577    2021-07-26      1400       100.0    6.0
10  5555555   10451    2021-08-13      5084         0.0    0.0
11  5555555    4864    2021-06-30      4581         8.0    1.0
12  5555555   13841    2021-05-27      3439       100.0    6.0
13  5555555   10388    2021-07-20      3031         0.0    0.0
14  5555555    3948    2021-06-02      9981       100.0    6.0
15  5555555   10195    2021-07-17      6641        26.0    3.0
16  5555555    4659    2021-05-29      1483       100.0    6.0
17  5555555   15407    2021-07-04      6177        82.0    5.0
18  5555555    6018    2021-07-26       959         9.0    2.0
19  5555555    7433    2021-05-31      5476        70.0    5.0
20  6666666    1145    2021-06-13      6237       100.0    6.0
21  6666666     158    2021-07-14      2602        12.0    2.0
22  6666666   14670    2021-07-31      7619       100.0    6.0
23  6666666    8424    2021-06-08      6752        25.0    3.0
24  6666666    8693    2021-05-30      7827        10.0    2.0
25  6666666    4676    2021-08-15      2703        70.0    5.0
26  6666666   13088    2021-06-14      4857        96.0    6.0
27  6666666    6836    2021-07-13      5555        80.0    5.0
28  6666666   15846    2021-08-17      7039         0.0    0.0
29  6666666     749    2021-08-18      1364        13.0    2.0
30  6666666    8202    2021-08-11      7814       100.0    6.0
31  6666666   13916    2021-06-19      7515        38.0    4.0
32  6666666    2051    2021-06-26      4771        93.0    5.0
33  6666666   15936    2021-05-02      1376       100.0    6.0
34  6666666    1059    2021-07-12      7456         5.0    1.0
35  6666666   10530    2021-05-25      6978         6.0    1.0
36  6666666     326    2021-08-19      5341        80.0    5.0
37  6666666   11647    2021-07-13      3952         9.0    2.0
38  6666666     665    2021-08-17      2617        20.0    3.0
39  6666666    1210    2021-08-17       826        82.0    5.0
40  7777777    2083    2021-06-19      5356       100.0    6.0
41  7777777     717    2021-07-30      9199       100.0    6.0
42  7777777   15257    2021-08-03      1807        95.0    6.0
43  7777777   11314    2021-06-21      8278        75.0    5.0
44  7777777     776    2021-07-21      8857       100.0    6.0
45  7777777   11048    2021-08-17      4240         9.0    2.0
46  7777777    1699    2021-07-16      8457        94.0    5.0
47  7777777   12421    2021-08-14      7043         0.0    0.0
48  7777777    9375    2021-05-21      8756       100.0    6.0
49  7777777    9779    2021-08-19      9603        29.0    3.0
50  7777777   12271    2021-05-10      7772        21.0    3.0
51  7777777    9676    2021-08-14      3075        95.0    6.0
52  7777777   14294    2021-07-12      3269       100.0    6.0
53  7777777   11481    2021-08-17      2160        13.0    2.0
54  7777777   13082    2021-08-04      4348       100.0    6.0
55  7777777    6203    2021-06-11      4334        53.0    4.0
56  7777777    6314    2021-06-10      2148        12.0    2.0
57  7777777   12773    2021-05-28      5436       100.0    6.0
58  7777777    2256    2021-08-07      4357       100.0    6.0
59  7777777   14046    2021-08-12      1012         0.0    0.0

In [187]:
interactions.shape

(4196975, 6)

In [191]:
interactions_with_avatar = pd.concat(
    [
        interactions,
        interactions_new_avatar
    ],
    ignore_index=True
).sample(
    interactions_with_avatar.shape[0]
).reset_index(drop=True)

interactions_with_avatar.shape[0] - interactions.shape[0] == 60

True

In [192]:
interactions_with_avatar.to_csv("../data/kion_train/interactions_with_avatar.csv", index=False)
users_with_avatar.to_csv("../data/kion_train/users_with_avatar.csv", index=False)

In [193]:
users_features = create_features_dataset(
    Columns.User, interactions_with_avatar, users_with_avatar, features["users"]
)
display_df(users_features)

id      value feature
0        973171  age_25_34     age
1        962099  age_18_24     age
3        721985  age_45_54     age
4        704055  age_35_44     age
5       1037719  age_45_54     age
840189   191349          М     sex
840190   393868          М     sex
840197  5555555          М     sex
840198  6666666          М     sex
840199  7777777          Ж     sex

In [234]:
users_features["id"].unique().shape

(626278,)

In [194]:
items_features = create_features_dataset(
    Columns.Item, interactions_with_avatar, items, features["items"]
)
display_df(items_features)

id     value       feature
0      10711      film  content_type
1       2508      film  content_type
2      10716      film  content_type
3       7868      film  content_type
4      16268      film  content_type
15958   6443  германия     countries
15959   2367    россия     countries
15960  10632    россия     countries
15961   4538    россия     countries
15962   3206    россия     countries

In [254]:
users_mapping = dict(enumerate(interactions_with_avatar[Columns.User].unique()))
users_inv_mapping = {v: k for k, v in users_mapping.items()}
items_mapping = dict(enumerate(interactions_with_avatar[Columns.Item].unique()))
items_inv_mapping = {v: k for k, v in items_mapping.items()}

print(f"Count unique users: {len(users_mapping)}")
print(f"Count unique items: {len(items_mapping)}")

Count unique users: 809580
Count unique items: 14169


### Построение лучших моделей с валидации на всех данных

In [416]:
dataset = Dataset.construct(
    interactions_df=interactions_with_avatar,
    user_features_df=users_features,
    cat_user_features=features["users"],
    item_features_df=items_features,
    cat_item_features=features["items"]
)

In [418]:
model_als = ImplicitALSWrapperModel(
    model=AlternatingLeastSquares(
        factors=32, 
        random_state=42, 
        num_threads=16,
    ), 
    fit_features_together=True
)

model_lfm = LightFMWrapperModel(
    LightFM(
        no_components=16, 
        loss="warp", 
        random_state=42,
        learning_rate=0.05, 
        user_alpha=0.2,
        item_alpha=0.2,
    ),
    epochs=10,
    num_threads=16,
)

In [419]:
%%time

model_als.fit(dataset)

CPU times: user 7min 36s, sys: 18.5 s, total: 7min 54s
Wall time: 1min 37s


In [420]:
%%time

model_lfm.fit(dataset)

CPU times: user 6min 12s, sys: 133 ms, total: 6min 12s
Wall time: 53.1 s


### Реализация приближенного поиска соседей для 3-ёх добавленных аватаров

In [491]:
import nmslib

In [492]:
user_embeddings_als, item_embeddings_als = model_als.get_vectors()
user_embeddings_als.shape, item_embeddings_als.shape

((809580, 237), (14169, 237))

In [493]:
user_embeddings_lfm, item_embeddings_lfm = model_lfm.get_vectors(dataset)
user_embeddings_lfm_bias.shape, item_embeddings_lfm_bias.shape

((809580, 18), (14169, 18))

In [494]:
def augment_inner_product(factors):
    normed_factors = np.linalg.norm(factors, axis=1)
    max_norm = normed_factors.max()
    
    extra_dim = np.sqrt(max_norm ** 2 - normed_factors ** 2).reshape(-1, 1)
    augmented_factors = np.append(factors, extra_dim, axis=1)
    return max_norm, augmented_factors

In [495]:
_, augmented_item_embeddings_als = augment_inner_product(item_embeddings_als)
_, augmented_item_embeddings_lfm = augment_inner_product(item_embeddings_lfm)

In [496]:
extra_zero_als = np.zeros((user_embeddings_als.shape[0], 1))
extra_zero_lfm = np.zeros((user_embeddings_lfm.shape[0], 1))
augmented_user_embeddings_als = np.append(user_embeddings_als, extra_zero_als, axis=1)
augmented_user_embeddings_lfm = np.append(user_embeddings_lfm, extra_zero_lfm, axis=1)

In [497]:
augmented_user_embeddings_als.shape, augmented_user_embeddings_lfm.shape

((809580, 238), (809580, 19))

In [498]:
CONFIG_NN = {
    "K": 10,
    "space_name": 'negdotprod',
    "method": 'hnsw',
    "index_time_params": {
        "M": 64,
        "efConstruction": 1000,
        "indexThreadQty": 4,
    },
    "query_time_params": {
        "efSearch": 1000,
    },
}

In [499]:
# Intitialize the library, specify the space, the type of the vector and add data points for each models
index_als = nmslib.init(
    method=CONFIG_NN["method"], space=CONFIG_NN["space_name"], data_type=nmslib.DataType.DENSE_VECTOR
) 
index_als.addDataPointBatch(augmented_item_embeddings_als) 

index_lfm = nmslib.init(
    method=CONFIG_NN["method"], space=CONFIG_NN["space_name"], data_type=nmslib.DataType.DENSE_VECTOR
) 
index_lfm.addDataPointBatch(augmented_item_embeddings_lfm) 

14169

In [500]:
# Create index for each model
index_als.createIndex(CONFIG_NN["index_time_params"]) 
index_lfm.createIndex(CONFIG_NN["index_time_params"]) 

In [501]:
# Setting query-time parameters
index_als.setQueryTimeParams(CONFIG_NN["query_time_params"])
index_lfm.setQueryTimeParams(CONFIG_NN["query_time_params"])

In [502]:
def get_reco_item(index, u_emb, user, users_inv_mapping, items_mapping, k=10):
    avatar_idx = users_inv_mapping[user]
    items_idx = index.knnQuery(u_emb[avatar_idx], k=k,)[0].tolist()
    return [items_mapping[idx] for idx in items_idx]

In [503]:
%%time 

models = ["als", "lfm"]
index_with_embs = [(index_als, augmented_user_embeddings_als), (index_lfm, augmented_user_embeddings_lfm)]

result = {
    "avatar": [],
    "model": [],
    "reco": [],
}
for avatar in [5555555, 6666666, 7777777]:
    for idx, (index, u_emb) in enumerate(index_with_embs):
        reco = get_reco_item(index, u_emb, avatar, users_inv_mapping, items_mapping, k=CONFIG_NN["K"])
        
        result["avatar"].append(avatar)
        result["model"].append(models[idx])
        result["reco"].append(reco)

CPU times: user 7.76 ms, sys: 0 ns, total: 7.76 ms
Wall time: 14.5 ms


In [504]:
df_result = pd.DataFrame(result).explode("reco")
df_result

avatar model   reco
0  5555555   als   9728
0  5555555   als   2280
0  5555555   als  13865
0  5555555   als   9254
0  5555555   als  10195
0  5555555   als  15443
0  5555555   als  10451
0  5555555   als    879
0  5555555   als   1577
0  5555555   als   4659
1  5555555   lfm  10440
1  5555555   lfm  15297
1  5555555   lfm  13865
1  5555555   lfm   4151
1  5555555   lfm   9728
1  5555555   lfm   3734
1  5555555   lfm   4880
1  5555555   lfm   2657
1  5555555   lfm   8636
1  5555555   lfm   4740
2  6666666   als   4676
2  6666666   als   5061
2  6666666   als   4088
2  6666666   als    602
2  6666666   als   9880
2  6666666   als  15141
2  6666666   als    599
2  6666666   als   9728
2  6666666   als  11237
2  6666666   als   1819
3  6666666   lfm  10440
3  6666666   lfm  15297
3  6666666   lfm  13865
3  6666666   lfm   4151
3  6666666   lfm   9728
3  6666666   lfm   3734
3  6666666   lfm   4880
3  6666666   lfm   2657
3  6666666   lfm   8636
3  6666666   lfm   4740
4  7777777   als  15297
4  7777777   als  10275
4  7777777   als   1541
4  7777777   als   2929
4  7777777   als  10725
4  7777777   als  12629
4  7777777   als    776
4  7777777   als  10448
4  7777777   als  13045
4  7777777   als   1687
5  7777777   lfm  10440
5  7777777   lfm  15297
5  7777777   lfm  13865
5  7777777   lfm   4151
5  7777777   lfm   9728
5  7777777   lfm   3734
5  7777777   lfm   4880
5  7777777   lfm   2657
5  7777777   lfm   8636
5  7777777   lfm   4740

In [505]:
df_result = df_result.merge(
    items[["item_id", "content_type", "genres", "countries"]], 
    left_on=["reco"],
    right_on=["item_id"], 
    how="left"
)
df_result

avatar model   reco  item_id content_type  \
0   5555555   als   9728     9728         film   
1   5555555   als   2280     2280       series   
2   5555555   als  13865    13865         film   
3   5555555   als   9254     9254       series   
4   5555555   als  10195    10195       series   
5   5555555   als  15443    15443       series   
6   5555555   als  10451    10451       series   
7   5555555   als    879      879       series   
8   5555555   als   1577     1577       series   
9   5555555   als   4659     4659       series   
10  5555555   lfm  10440    10440       series   
11  5555555   lfm  15297    15297       series   
12  5555555   lfm  13865    13865         film   
13  5555555   lfm   4151     4151       series   
14  5555555   lfm   9728     9728         film   
15  5555555   lfm   3734     3734         film   
16  5555555   lfm   4880     4880       series   
17  5555555   lfm   2657     2657       series   
18  5555555   lfm   8636     8636         film   
19  5555555   lfm   4740     4740         film   
20  6666666   als   4676     4676         film   
21  6666666   als   5061     5061         film   
22  6666666   als   4088     4088         film   
23  6666666   als    602      602         film   
24  6666666   als   9880     9880       series   
25  6666666   als  15141    15141         film   
26  6666666   als    599      599         film   
27  6666666   als   9728     9728         film   
28  6666666   als  11237    11237         film   
29  6666666   als   1819     1819         film   
30  6666666   lfm  10440    10440       series   
31  6666666   lfm  15297    15297       series   
32  6666666   lfm  13865    13865         film   
33  6666666   lfm   4151     4151       series   
34  6666666   lfm   9728     9728         film   
35  6666666   lfm   3734     3734         film   
36  6666666   lfm   4880     4880       series   
37  6666666   lfm   2657     2657       series   
38  6666666   lfm   8636     8636         film   
39  6666666   lfm   4740     4740         film   
40  7777777   als  15297    15297       series   
41  7777777   als  10275    10275         film   
42  7777777   als   1541     1541         film   
43  7777777   als   2929     2929         film   
44  7777777   als  10725    10725         film   
45  7777777   als  12629    12629         film   
46  7777777   als    776      776         film   
47  7777777   als  10448    10448       series   
48  7777777   als  13045    13045         film   
49  7777777   als   1687     1687         film   
50  7777777   lfm  10440    10440       series   
51  7777777   lfm  15297    15297       series   
52  7777777   lfm  13865    13865         film   
53  7777777   lfm   4151     4151       series   
54  7777777   lfm   9728     9728         film   
55  7777777   lfm   3734     3734         film   
56  7777777   lfm   4880     4880       series   
57  7777777   lfm   2657     2657       series   
58  7777777   lfm   8636     8636         film   
59  7777777   lfm   4740     4740         film   

                                               genres              countries  
0                                 [боевики, триллеры]  [великобритания, сша]  
1                    [боевики, историческое, военные]     [россия, беларусь]  
2                       [драмы, военные, приключения]               [россия]  
3                                  [боевики, военные]             [беларусь]  
4                           [боевики, драмы, русские]               [россия]  
5                  [драмы, боевики, русские, военные]               [россия]  
6                                  [боевики, русские]               [россия]  
7                                  [боевики, русские]               [россия]  
8                                  [боевики, русские]               [россия]  
9                                  [боевики, русские]               [россия]  
10                              [триллеры, детективы]               [россия]  
11    

**avatar 1**
- Возраст от 45-54, доход 60-90, пол мужской, смотрим русские сериалы боевики (**id 5555555**)

In [506]:
df_result[df_result["avatar"] == 5555555][["model", "reco", "content_type", "genres", "countries"]]

model   reco content_type                              genres  \
0    als   9728         film                 [боевики, триллеры]   
1    als   2280       series    [боевики, историческое, военные]   
2    als  13865         film       [драмы, военные, приключения]   
3    als   9254       series                  [боевики, военные]   
4    als  10195       series           [боевики, драмы, русские]   
5    als  15443       series  [драмы, боевики, русские, военные]   
6    als  10451       series                  [боевики, русские]   
7    als    879       series                  [боевики, русские]   
8    als   1577       series                  [боевики, русские]   
9    als   4659       series                  [боевики, русские]   
10   lfm  10440       series               [триллеры, детективы]   
11   lfm  15297       series                  [драмы, мелодрамы]   
12   lfm  13865         film       [драмы, военные, приключения]   
13   lfm   4151       series                           [комедии]   
14   lfm   9728         film                 [боевики, триллеры]   
15   lfm   3734         film                           [комедии]   
16   lfm   4880       series                           [комедии]   
17   lfm   2657       series                   [драмы, триллеры]   
18   lfm   8636         film                      [драмы, спорт]   
19   lfm   4740         film                    [документальное]   

                countries  
0   [великобритания, сша]  
1      [россия, беларусь]  
2                [россия]  
3              [беларусь]  
4                [россия]  
5                [россия]  
6                [россия]  
7                [россия]  
8                [россия]  
9                [россия]  
10               [россия]  
11               [россия]  
12               [россия]  
13               [россия]  
14  [великобритания, сша]  
15               [россия]  
16               [россия]  
17               [россия]  
18               [россия]  
19               [россия]

**avatar 2**
- Возраст от 18-24, доход 20-40, пол мужской, смотрит фильмы жанра комедия страна любая кроме России (**id 6666666**)

In [507]:
df_result[df_result["avatar"] == 6666666][["model", "reco", "content_type", "genres", "countries"]]

model   reco content_type  \
20   als   4676         film   
21   als   5061         film   
22   als   4088         film   
23   als    602         film   
24   als   9880       series   
25   als  15141         film   
26   als    599         film   
27   als   9728         film   
28   als  11237         film   
29   als   1819         film   
30   lfm  10440       series   
31   lfm  15297       series   
32   lfm  13865         film   
33   lfm   4151       series   
34   lfm   9728         film   
35   lfm   3734         film   
36   lfm   4880       series   
37   lfm   2657       series   
38   lfm   8636         film   
39   lfm   4740         film   

                                               genres              countries  
20                               [мелодрамы, комедии]                 [кипр]  
21                             [приключения, комедии]                 [кипр]  
22                                   [драмы, комедии]                 [кипр]  
23                              [мультфильм, комедии]                 [кипр]  
24  [для детей, приключения, зарубежные, семейное,...                 [кипр]  
25                             [зарубежные, триллеры]                 [кипр]  
26                                       [мультфильм]                 [кипр]  
27                                [боевики, триллеры]  [великобритания, сша]  
28                                          [комедии]               [россия]  
29                                          [комедии]               [россия]  
30                              [триллеры, детективы]               [россия]  
31                                 [драмы, мелодрамы]               [россия]  
32                      [драмы, военные, приключения]               [россия]  
33                                          [комедии]               [россия]  
34                                [боевики, триллеры]  [великобритания, сша]  
35                                          [комедии]               [россия]  
36                                          [комедии]               [россия]  
37                                  [драмы, триллеры]               [россия]  
38                                     [драмы, спорт]               [россия]  
39                                   [документальное]               [россия]

**avatar 3**
 - Возраст от 35-44, доход 40-60, пол женский, смотрим и фильмы и сериалы жанра мелодрамы неважна какая страна (**id 7777777**)

In [508]:
df_result[df_result["avatar"] == 7777777][["model", "reco", "content_type", "genres", "countries"]]

model   reco content_type  \
40   als  15297       series   
41   als  10275         film   
42   als   1541         film   
43   als   2929         film   
44   als  10725         film   
45   als  12629         film   
46   als    776         film   
47   als  10448       series   
48   als  13045         film   
49   als   1687         film   
50   lfm  10440       series   
51   lfm  15297       series   
52   lfm  13865         film   
53   lfm   4151       series   
54   lfm   9728         film   
55   lfm   3734         film   
56   lfm   4880       series   
57   lfm   2657       series   
58   lfm   8636         film   
59   lfm   4740         film   

                                               genres              countries  
40                                 [драмы, мелодрамы]               [россия]  
41  [приключения, мюзиклы, мелодрамы, семейное, ко...                 [ссср]  
42                      [мелодрамы, мюзиклы, комедии]                 [ссср]  
43                      [мелодрамы, мюзиклы, комедии]                 [ссср]  
44                      [мелодрамы, мюзиклы, комедии]                 [ссср]  
45                      [мелодрамы, мюзиклы, комедии]                 [ссср]  
46                          [драмы, спорт, мелодрамы]                 [ссср]  
47                        [драмы, мюзиклы, мелодрамы]                 [ссср]  
48                        [драмы, мюзиклы, мелодрамы]                 [ссср]  
49                        [драмы, мюзиклы, мелодрамы]                 [ссср]  
50                              [триллеры, детективы]               [россия]  
51                                 [драмы, мелодрамы]               [россия]  
52                      [драмы, военные, приключения]               [россия]  
53                                          [комедии]               [россия]  
54                                [боевики, триллеры]  [великобритания, сша]  
55                                          [комедии]               [россия]  
56                                          [комедии]               [россия]  
57                                  [драмы, триллеры]               [россия]  
58                                     [драмы, спорт]               [россия]  
59                                   [документальное]               [россия]

- als имеет хорошие рекомендации, но тут может получиться зациклинность, что в реках будут выпадать только боевики
- lfm модели имеют хужие предсказания, но зато пользователю будут попадаться новые item из других категорий

**Возникла проблема, что LigthFM возвращает постоянно одни и те же рекомендации**, не удалось оддебажить. Если получится найти ошибку, буду благодарен!

### Сохраним вектора моделей для рекомендаций в сервисе

In [485]:
with open("../data/hw_4/als_users.npy", "wb") as file:
    np.save(file, augmented_user_embeddings_als)
    
with open("../data/hw_4/als_items.npy", "wb") as file:
    np.save(file, augmented_item_embeddings_als)

In [486]:
with open("../data/hw_4/lfm_users.npy", "wb") as file:
    np.save(file, augmented_user_embeddings_lfm)
    
with open("../data/hw_4/lfm_items.npy", "wb") as file:
    np.save(file, augmented_item_embeddings_lfm)

In [487]:
np.load("../data/hw_4/als_users.npy").shape

(809580, 238)

In [488]:
np.load("../data/hw_4/als_items.npy").shape

(14169, 238)

In [489]:
np.load("../data/hw_4/lfm_users.npy").shape

(809580, 19)

In [490]:
np.load("../data/hw_4/lfm_items.npy").shape

(14169, 19)

# Для холодных пользователей будут возвращаться популярное аналогично ДЗ-3, т.к. получился хороший результат популярного 0,08 по метрике MAP@10